### PR TITLE
audit: forbid nested dependencies in depends_on declarations

### LINKED ISSUE #40264: Concretizer internal error with a spec that previously worked
### Steps to reproduce

```console
./bin/spack spec -I xsdk-examples@0.4.0
```

### Error message

```console
==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable, errors are:
```

Debug output:
```console
>./bin/spack -d spec -I xsdk-examples@0.4.0
==> [2023-09-30-13:11:05.949225] Imported spec from built-in commands
==> [2023-09-30-13:11:05.949588] Imported spec from built-in commands
==> [2023-09-30-13:11:05.950212] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/config.yaml
==> [2023-09-30-13:11:05.963093] DATABASE LOCK TIMEOUT: 60s
==> [2023-09-30-13:11:05.963133] PACKAGE LOCK TIMEOUT: No timeout
==> [2023-09-30-13:11:05.965687] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/bootstrap.yaml
==> [2023-09-30-13:11:05.990111] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/repos.yaml
==> [2023-09-30-13:11:05.992437] [BOOTSTRAP CONFIG SCOPE] name=_builtin
==> [2023-09-30-13:11:05.992654] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/bootstrap.yaml
==> [2023-09-30-13:11:05.996474] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=/Users/dobrev1/spack/etc/spack/defaults
==> [2023-09-30-13:11:05.996501] [BOOTSTRAP CONFIG SCOPE] name=defaults/darwin, path=/Users/dobrev1/spack/etc/spack/defaults/darwin
==> [2023-09-30-13:11:05.996523] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=/Users/dobrev1/.spack/bootstrap/config
==> [2023-09-30-13:11:05.996540] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/darwin, path=/Users/dobrev1/.spack/bootstrap/config/darwin
==> [2023-09-30-13:11:05.997031] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/config.yaml
==> [2023-09-30-13:11:06.008945] Reading config from file /Users/dobrev1/.spack/bootstrap/config/darwin/config.yaml
==> [2023-09-30-13:11:06.014515] DATABASE LOCK TIMEOUT: 60s
==> [2023-09-30-13:11:06.014562] PACKAGE LOCK TIMEOUT: No timeout
==> [2023-09-30-13:11:06.014733] Reading config from file /Users/dobrev1/.spack/bootstrap/config/darwin/compilers.yaml
==> [2023-09-30-13:11:06.044687] [BOOTSTRAP ROOT SPEC] gnupg@2.3: %apple-clang target=aarch64
==> [2023-09-30-13:11:06.061290] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %apple-clang target=aarch64
==> [2023-09-30-13:11:06.061419] [BOOTSTRAP MODULE clingo] Try importing from Python
==> [2023-09-30-13:11:06.061689] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/bootstrap.yaml
==> [2023-09-30-13:11:06.176991] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/packages.yaml
==> [2023-09-30-13:11:06.194389] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/darwin/packages.yaml
==> [2023-09-30-13:11:06.222105] [BOOTSTRAP MODULE clingo] The installed spec "clingo-bootstrap@spack+python %apple-clang target=aarch64 ^python@3.9/p2cyqcsow6k6prfryoqb7usv27hhofuq" provides the "clingo" Python module
==> [2023-09-30-13:11:06.222535] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/concretizer.yaml
==> [2023-09-30-13:11:06.227132] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/repos.yaml
==> [2023-09-30-13:11:06.278689] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/packages.yaml
==> [2023-09-30-13:11:06.296195] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/darwin/packages.yaml
==> [2023-09-30-13:11:06.303387] Reading config from file /Users/dobrev1/spack/etc/spack/packages.yaml
==> [2023-09-30-13:11:06.331952] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/config.yaml
==> [2023-09-30-13:11:06.344053] Reading config from file /Users/dobrev1/spack/etc/spack/defaults/mirrors.yaml
==> [2023-09-30-13:11:08.636503] Reading config from file /Users/dobrev1/spack/etc/spack/compilers.yaml
==> [2023-09-30-13:11:22.908916] SpackError: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable, errors are:
==> [2023-09-30-13:11:22.909025] Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable, errors are:
Traceback (most recent call last):
  File "/Users/dobrev1/spack/lib/spack/spack/cmd/__init__.py", line 218, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/Users/dobrev1/spack/lib/spack/spack/spec.py", line 2959, in concretize
    self._new_concretize(tests)
  File "/Users/dobrev1/spack/lib/spack/spack/spec.py", line 2932, in _new_concretize
    result.raise_if_unsat()
  File "/Users/dobrev1/spack/lib/spack/spack/solver/asp.py", line 506, in raise_if_unsat
    raise InternalConcretizerError(constraints, conflicts=conflicts)
spack.solver.asp.InternalConcretizerError: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable, errors are:

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/dobrev1/spack/lib/spack/spack/main.py", line 1022, in main
    return _main(argv)
  File "/Users/dobrev1/spack/lib/spack/spack/main.py", line 977, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/Users/dobrev1/spack/lib/spack/spack/main.py", line 1005, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/Users/dobrev1/spack/lib/spack/spack/main.py", line 646, in _invoke_command
    return_val = command(parser, args)
  File "/Users/dobrev1/spack/lib/spack/spack/cmd/spec.py", line 101, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "/Users/dobrev1/spack/lib/spack/spack/cmd/__init__.py", line 232, in parse_specs
    raise spack.error.SpackError(msg) from e
spack.error.SpackError: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable, errors are:
```

### Information on your system

```console
>./bin/spack debug report
* **Spack:** 0.21.0.dev0 (08a9345fccd6c3cb9c574101e9b98abc1e2c4131)
* **Python:** 3.9.6
* **Platform:** darwin-ventura-m2
* **Concretizer:** clingo
```
The default configured compiler is `apple-clang@14.0.3`.

⚠️ This issue is not specific to this machine -- it happens on other machines as well, e.g. on Linux with gcc 11.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### LINKED ISSUE #41421: Audit failure: allows spec expressions that cause concretization failures
### Steps to reproduce

(https://github.com/spack/spack/issues/41405 reports the same error, although I think the issue is different because it's about a spec requested from the cmd line)

Create a package `foo` with

```
depends_on("mfem+lapack+netcdf+mpi ^netcdf-c~shared")
```

then

```
spack audit packages foo
```

reports no problems, but `spack spec foo` reports:

```
==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    foo... is unsatisfiable
```

the problem is `^netcdf-c~shared`: we don't explicitly forbid this, but whatever this expands to in ASP causes the solve to fail.

### Error message

(see reproducer)

### Information on your system

This was for spack 4c2734fe14be603c4d3028ad10c44fa3f72b02a2 on a Linux (RHEL) system.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by alalazo**
fixes #41421 
fixes #40264

This PR adds a new audit that is checked in CI. The audit prevent declarations like:
```
depends_on("a ^b ^c")
```
and encourages rewriting as:
```
depends_on("a")
depends_on("b")
depends_on("c")
```

### CODE REVIEW

**Comment by alalazo** [comment]
This:
- https://github.com/spack/spack/actions/runs/7101293444/job/19329103520?pr=41428

shows the audit is actually catching issues.

File: var/spack/repos/builtin/packages/exago/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/exago/package.py:134]: Not really sure what this was meant to do, so I simplified it. Also, kind of weird to use "format"... with literal strings as arguments.

Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: @cameronrutherford @pelesh @ryandanehy Let me know if this change is fine, or if you need a different semantic.

**Comment by alalazo** [comment]
@spackbot maintainers

**Review by snehring — APPROVED** [review]
Looks good to me. Personally I think the original way gets the point across more clearly, but ultimately it doesn't matter if it's all being concretized together.

**Review by balay — DISMISSED** [review]
Thanks! This fixes the following issue for me:
```
balay@petsc-gpu-02:/scratch/balay/spack$ ./bin/spack spec xsdk-examples@0.4.0
==> Error: Spack concretizer internal error. Please submit a bug report and include the command, environment if applicable and the following error message.
    xsdk-examples@0.4.0 is unsatisfiable
```

File: lib/spack/spack/audit.py
Comment (scheibelp) [inline lib/spack/spack/audit.py:758]: Given the name of the function and the docstring, I wouldn't have expected to find this check in it (the problem doesn't have to do with variants, and the dependency can be known) - I think updating the docstring would be helpful (IMO it could just list out the individual things that are checked for).

Also, does it make sense to put this before 

```
            if spack.repo.PATH.is_virtual(dependency_name):
                continue
```

? Virtuals can't have dependencies, so I'd assume we could also skip this check for virtuals in the same way we would skip the others.

File: var/spack/repos/builtin/packages/exago/package.py
Comment (scheibelp) [reply var/spack/repos/builtin/packages/exago/package.py:134]: It looks like this was trying to propagate the `build_type` and that this is more concerned with adding a dependency on `ginkgo` - expanding the second one:

```
depends_on("hiop+ginkgo ^ginkgo build_type=Debug",
  when="+hiop ^hiop+ginkgo build_type=RelWithDebInfo")
```

so another interpretation could be:

```
with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWIthDebInfo)
```

In words: "I need ginkgo when I need hiop and hiop needs ginkgo; also build ginkgo in debug mode if hiop was built in debug mode"

(there's already a dependency on `hiop` added up around line 106)

**Review by sebastiangrimberg — APPROVED** [review]

**Review by scheibelp — COMMENTED** [review]
I agree that the `exago` constraints are confusing: I added my own interpretation. I also have a suggestion for docs.

File: var/spack/repos/builtin/packages/exago/package.py
Comment (ryandanehy) [reply var/spack/repos/builtin/packages/exago/package.py:134]: @jaelynlitz and I are in agreement with @scheibelp interpretations we already use the simple depends on earlier in the code when we constrain hiop version number lines 136-140. This snippet is trying to propagate build type to packages. So we like @scheibelp suggestion of

```
with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Release", when="^hiop build_type=Release)
```

```
with when("+hiop ^hiop+ginkgo"):
  depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWIthDebInfo)
```
Note: we still have to test to make sure.

Comment (jaelynlitz) [reply var/spack/repos/builtin/packages/exago/package.py:134]: After some fiddling locally -  this seems to give the correct behavior on our side, not sure if this will pass on the audit

```
     with when("+hiop build_type=Release ^hiop+ginkgo"):                         
       depends_on("hiop build_type=Release")                                     
       depends_on("ginkgo build_type=Release", when="^hiop build_type=Release")  
                                                                                 
     with when("+hiop build_type=Debug ^hiop+ginkgo"):                           
       depends_on("hiop build_type=RelWithDebInfo")                                                                                                                                
       depends_on("ginkgo build_type=Debug", when="^hiop build_type=RelWithDebInfo")
```

File: lib/spack/spack/audit.py
Comment (alalazo) [reply lib/spack/spack/audit.py:758]: > Also, does it make sense to put this before ...

In my opinion yes, since this will catch issues with nested dependencies on virtuals - instead of silently dropping the dependency.

Comment (alalazo) [reply lib/spack/spack/audit.py:758]: Fixed in https://github.com/spack/spack/pull/41428/commits/e7c2cae13623e2d754c36702ff0a5fd1c1d34b45

File: var/spack/repos/builtin/packages/exago/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: See https://github.com/spack/spack/pull/41428/commits/68bae59b80b6d33832bd3e6bcf973abc8b239a55 I avoided things like:
```
depends_on("foo", when="^foo")
```
cause they are tautological and might add nondeterminism by giving clingo the "choice" to put an extra node. I tried to propagate the `build_type` like you do here, though I think this is better done using configuration.

Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: `exago` is failing to concretize in pipelines. I'm looking at what is the issue, but cannot reproduce locally so far.

**Comment by alalazo** [comment]
@spackbot run pipeline

File: var/spack/repos/builtin/packages/exago/package.py
Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: Apparently on `develop` the spec `exago+hiop` depends on `hiop~ginkgo`, while this PR forces `hiop+ginkgo`. Is it expected to depend on `hiop~ginkgo` on `develop` when `+hiop`?

Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: It took a while, but this is failing because of this hidden feature enabled only in CI https://github.com/spack/spack/pull/39611 Explanation:
1. In CI we have the environment variable SPACK_CONCRETIZER_REQUIRE_CHECKSUM set
2. `hiop+ginkgo` depends on a branch that is not checksummed of `ginkgo`
3. CI fails to concretize

On `develop` the concretization gives `exago ^hiop~ginkgo`

Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: @jaelynlitz @ryandanehy There are a couple of points that are still not clear to me: 
1. Does `exago+hiop` require `hiop+ginkgo` or is `hiop~ginkgo` fine too?
2. If `hiop~ginkgo` is fine, do I understand correctly from https://github.com/spack/spack/pull/41428#discussion_r1416305778 that you want to forward the build types only if `hiop+ginkgo` is used?

**Review by scheibelp — DISMISSED** [review]
I'm approving the auditing logic specifically. I'll leave it up to you to resolve the `exago` constraints and merge when satisfied with them (although I'm happy to provide additional suggestions if desired).

File: var/spack/repos/builtin/packages/exago/package.py
Comment (jaelynlitz) [reply var/spack/repos/builtin/packages/exago/package.py:134]: 1. I believe `hiop~ginkgo` is viable, but yes `exago+hiop+ginkgo` only works with the `glu_experimental` branch of ginkgo. 
2. Our goal is to in general propagate the same or similar `build_type` down to our dependencies and their dependencies. Like in this loop of packages [L112-125](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/exago/package.py#L120). If that syntax looks okay, we could probably just move `hiop+ginkgo` to be included there?

Comment (alalazo) [reply var/spack/repos/builtin/packages/exago/package.py:134]: @jaelynlitz I think I implemented the semantic you asked for (forward build type, `exago+hiop` can work with both `hiop~ginkgo` and `hiop+ginkgo`). I'm going to merge this PR, but ping me if you think we need further changes on `exago`.

Comment (cameronrutherford) [reply var/spack/repos/builtin/packages/exago/package.py:134]: https://github.com/pnnl/ExaGO/issues/50 - tracks us upgrading from the old ginkgo version. Thanks all for the help, and I think this change works well

**Pull request merged**
